---
title: Project Report
subtitle: {{ project.ngi_name }}_project_summary
date: {{ report_info.report_date }}
support_email: {{ report_info.support_email }}
---

# Project Information

User Project Name
:   {{ project.user_ID }}

NGI Project Name
:   {{ project.ngi_name }}

NGI Project ID
:   {{ project.ngi_id }}

{% if project.ngi_facility -%}
NGI Facility
:   {{ project.ngi_facility }}
{%- endif %}

User Contact
:   [{{ project.contact }}](mailto:{{ project.contact }})

NGI Application Type
:   {{ project.application }} _({% if project.best_practice %}including best practice analysis{% else %}no best practice analysis{% endif %})_

{% if project.num_samples -%}
Submitted samples
:   {{ project.num_samples }} sample{% if project.num_samples > 1 %}s{% endif %}
{%- endif %}

{% if project.num_lanes -%}
Ordered Flow cells
:   {{ project.num_lanes }} flow cell{% if project.num_lanes > 1 %}s{% endif %}
{%- endif %}

Order Dates
:   {{ report_info.dates }}

{% if project.reference.genome -%}
Reference Genome
:   {% if project.reference.organism -%}{{ project.reference.organism}} - {% endif %}{{ project.reference.genome }}
{%- endif %}

Report generated by
:   {{ report_info.signature }}, {{ report_info.report_date }}

NGI is an accredited facility. However, the workflows used for this project is not under accreditation.

# Methods

{% if project.library_construction -%}
### Library construction
{{ project.library_construction }}
{%- endif %}

{% if report_info.sequencing_methods -%}
### Sequencing
{{ report_info.sequencing_methods }}
{%- endif %}

### Data Processing
To ensure that all sequenced data meets our guarantee of data quality and quantity,
a number of standardised bioinformatics quality control checks are performed before
delivery. These include checking the yield, sequence read quality and average read length.


# Sample Information
{% if not project.samples %}
No sample information to be displayed.
{% else %}
NGI ID | User ID | {{ project.samples_unit }}
-------|---------|----------
{% for sample in project.samples.values()|sort(attribute='ngi_id') -%}
{{ sample.ngi_id }} | `{{ sample.customer_name }}` | {{ sample.total_reads }}
{% endfor %}


Below you can find an explanation of the header column used in the table.

{{ tables.sample_info }}
{% endif %}


# Library Information
{% if not project.samples %}
No library information to be displayed.
{% else %}
NGI ID | Index | Avg. FS (bp) | Lib. QC
-------|-------|--------------|--------
{% for sample in project.samples.values()|sort(attribute='ngi_id') -%}
{% if sample.preps -%}
{% for prep in sample.preps.values() -%}
{{ sample.ngi_id }} | `{{ prep.barcode }}` | {{ prep.avg_size }} | {{ prep.qc_status }}
{% endfor -%}
{% endif -%}
{%- endfor %}


Below you can find an explanation of the header column used in the table.

{{ tables.library_info }}
{% endif %}


# Flow cell Information
{% if project.missing_fc %}
No flow cell information to be displayed.
{% else %}
Date | Flow cell | Reads (M) | N50
-----|----------|-------|----
{% for fc in project.flowcells.values()|sort(attribute='date') -%}
{{ fc.date }} | `{{ fc.run_name }}` | {{ fc.total_reads }} | {{ fc.n50 }}
{% endfor %}

Below you can find an explanation of the header column used in the table.

{{ tables.lanes_info }}
{% endif %}

# Additions to, deviations or exclusions from the accredited method(s)

None have been reported for this project.

{% if project.aborted_samples %}
# Aborted/Not Sequenced samples

NGI ID | User ID | Status
-------|---------|-------
{% for sample, info in project.aborted_samples.items() -%}
{{ sample }} | {{ info.user_id }} | {{ info.status }}
{% endfor -%}
{% endif %}

# General Information

The results apply to the sample(s) as received.

## Samples and/or libraries that have failed RC (Reception control) and/or Lib. QC (Library quality control)

In cases where samples and/or libraries have failed the QC, the user is always consulted regarding how to proceed. If the user wishes to proceed to sequence the failed samples and/or libraries, NGI bears no responsibility regarding the quality and number of reads of the sequenced sample data.

## Data access at DDS

Data have been uploaded to the **Data Delivery System**
(DDS, [https://delivery.scilifelab.se](https://delivery.scilifelab.se)).
More details can be found on the following links

- [DDS user guide](https://ngisweden.scilifelab.se/resources/data-delivery-dds/)
- [Youtube tutorial for DDS](https://youtu.be/Mk5nhClsfTA)
- [DDS official document](https://scilifelabdatacentre.github.io/dds_cli/)

## Acknowledgements

Services provided by NGI must be acknowledged in all scientific dissemination activities,
such as publications, presentations, posters, etc., using the following statement:

> The authors acknowledge support from the National Genomics Infrastructure in {% if project.ngi_facility -%} {{ project.ngi_facility }} {%- endif %}
> funded by Science for Life Laboratory, the Knut and Alice Wallenberg Foundation and the Swedish Research Council,
> and NAISS/Uppsala Multidisciplinary Center for Advanced Computational Science for assistance with massively parallel sequencing
> and access to the UPPMAX computational infrastructure

This acknowledgement is used for reporting purposes by the NGI and is critical for the future funding of the facility.

# Further Help
If you have any queries, please get in touch at
[{{ report_info.support_email }}](mailto:{{ report_info.support_email }}).


Postal address | Visiting address | Shipping address
:-------|:---------|:-------
NGI Stockholm<br>NGI-Alfa3<br>Science for Life Laboratory<br>Box 1031<br>17121 Solna, Sweden | SciLifeLab Stockholm<br>Science for Life Laboratory,<br>Tomtebodavägen 23A,<br>17165 Solna, Sweden | Att: Mattias Ormestad<br>NGI-SciLifeLab<br>Tomtebodavägen 23B,<br>17165 Solna, Sweden